### PR TITLE
[SPARK-29816][MLLIB] Missing persist in mllib.evaluation.BinaryClassificationMetrics.recallByThreshold()

### PR DESCRIPTION
**Pull request by amanomer**
### What changes were proposed in this pull request?
Use `persist` on RDD which is used for more than one action.

### Why are the changes needed?
The rdd `scoreAndLabels.combineByKey` is used by two actions: `sortByKey` and `count`, so it needs to be persisted.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Tested manually.

### CODE REVIEW

**Comment by amanomer** [comment]
@MaxGekk Kindly review this PR.

**Comment by srowen** [comment]
There's no reason this has to be like 5 JIRAs and PRs. Just combine them into one, please.

**Comment by amanomer** [comment]
Okay.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala [hunk 1]
```diff
@@ -165,13 +165,15 @@ class BinaryClassificationMetrics @Since("3.0.0") (
...
-    val counts = scoreLabelsWeight.combineByKey(
+    val binnedWeights = scoreLabelsWeight.combineByKey(
       createCombiner = (labelAndWeight: (Double, Double)) =>
         new BinaryLabelCounter(0.0, 0.0) += (labelAndWeight._1, labelAndWeight._2),
       mergeValue = (c: BinaryLabelCounter, labelAndWeight: (Double, Double)) =>
         c += (labelAndWeight._1, labelAndWeight._2),
       mergeCombiners = (c1: BinaryLabelCounter, c2: BinaryLabelCounter) => c1 += c2
-    ).sortByKey(ascending = false)
+    )
+    binnedWeights.persist()
```
Comment (MaxGekk) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala]: Could you explain why do you persist `binnedWeights` but not `counts`? As I can see `binnedWeights` is used only once ... or I missed something?

Comment (amanomer) [reply mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala]: Rdd `scoreAndLabels.combineByKey` is found to be used more than one time, as mentioned in [SPARK-29816](https://issues.apache.org/jira/browse/SPARK-29816).
I think, `counts` rdd is constructed over `scoreAndLabels.combineByKey` after action `sortByKey`. `counts` is indirectly dependent on `scoreAndLabels.combineByKey` i.e, `binnedWeights`?

**Comment by srowen** [comment]
Let's please merge this to https://github.com/apache/spark/pull/26454 after discussing some initial principles there.

**Pull request closed without merging**